samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -183,96 +152,100 @@ class KafkaCheckpointManager(
+  }
 
   /**
-   * Common code for reading both changelog partition mapping and change log
+   * Reads an entry from the checkpoint log and invokes the provided lambda on that entry.
    *
-   * @param entryType What type of entry to look for within the log key's
    * @param handleEntry Code to handle an entry in the log once it's found
    */
-  private def readLog(entryType:String, shouldHandleEntry: (KafkaCheckpointLogKey) => Boolean,
+  private def readLog(shouldHandleEntry: (KafkaCheckpointLogKey) => Boolean,
                       handleEntry: (ByteBuffer, KafkaCheckpointLogKey) => Unit): Unit = {
-    retryBackoff.run[Unit](
-      loop => {
-        val consumer = getConsumer()
-
-        val topicAndPartition = new TopicAndPartition(checkpointTopic, 0)
 
+    val UNKNOWN_OFFSET: String = "-1"
+    var attempts = 10
+    val POLL_TIMEOUT = 1000L
+
+    val ssp: SystemStreamPartition = new SystemStreamPartition(systemName, checkpointTopic, new Partition(0))
+    val systemConsumer = getSystemConsumer()
+    val partitionMetadata = getEarliestOffset(checkpointTopic, new Partition(0))
+    // offsets returned are strings
+    val newestOffset = if (partitionMetadata.getNewestOffset == null) UNKNOWN_OFFSET else partitionMetadata.getNewestOffset
+    val oldestOffset = partitionMetadata.getOldestOffset
+    systemConsumer.register(ssp, oldestOffset) // checkpoint stream should always be read from the beginning
+    systemConsumer.start()
+
+    var msgCount = 0
+    try {
+      val emptyEnvelops = util.Collections.emptyMap[SystemStreamPartition, java.util.List[IncomingMessageEnvelope]]
+      // convert offsets to long
+      var currentOffset = UNKNOWN_OFFSET.toLong
+      val newestOffsetLong = newestOffset.toLong
+      val sspToPoll = Collections.singleton(ssp)
+      while (currentOffset < newestOffsetLong) {
+
+        val envelopes: java.util.Map[SystemStreamPartition, java.util.List[IncomingMessageEnvelope]] =
         try {
-          var offset = startingOffset.getOrElse(getEarliestOffset(consumer, topicAndPartition))
-
-          info("Got offset %s for topic %s and partition 0. Attempting to fetch messages for %s." format(offset, checkpointTopic, entryType))
-
-          val latestOffset = getOffset(consumer, topicAndPartition, OffsetRequest.LatestTime)
-
-          info("Get latest offset %s for topic %s and partition 0." format(latestOffset, checkpointTopic))
-
-          if (offset < 0) {
-            info("Got offset 0 (no messages in %s) for topic %s and partition 0, so returning empty collection. If you expected the checkpoint topic to have messages, you're probably going to lose data." format (entryType, checkpointTopic))
-            return
+          systemConsumer.poll(sspToPoll, POLL_TIMEOUT)
+        } catch {
+          case e: Exception => {
+            // these exceptions are most likely intermediate
+            warn("Got %s exception while polling the consumer for checkpoints." format e.getMessage)
```

Comment:
should you print the entire stack trace for the exception instead of just the message?

Reply:
k